template/{% if use_git and dev_platform == 'GitHub' %}.github{% endif %}/workflows/release.yml

```
@@ -1,53 +1,74 @@
+  build:
+    runs-on: ubuntu-latest
+    # This ensures that the publish action only runs in the main repository
+    # rather than forks
+    # Environment is encouraged for increased security
+    environment: build
     steps:
-    - uses: actions/checkout@v4
-
-    - name: Set up Python ${{ matrix.python-version }}
-      uses: actions/setup-python@v5
-      with:
-        python-version: ${{ matrix.python-version }}
+      - name: Checkout
+        uses: actions/checkout@v4
+        with:
+          # This fetch element is only important if you are use SCM based
+          # versioning (that looks at git tags to gather the version)
+          fetch-depth: 100
+            # Need the tags so that setuptools-scm can form a valid version number
+      - name: Fetch git tags
+        run: git fetch origin 'refs/tags/*:refs/tags/*'
 
-    - name: Install hatch
-      uses: pypa/hatch@install
+      - name: Setup Python
+        uses: actions/setup-python@v5
+        with:
+          # You can modify what version of Python you want to use for your release
+          python-version: "3.11"
 
-    - name: Build package
-      run: hatch build
+      # Security recommends we should pin deps. Should we pin the workflow version?
+      - name: Install hatch
+        uses: pypa/hatch@install
 
-    # We rely on a trusted publisher configuration being present on PyPI,
-    # see https://docs.pypi.org/trusted-publishers/.
-    - name: Publish to PyPI
-      uses: pypa/gh-action-pypi-publish@release/v1
+      - name: Build package using Hatch
+        run: |
+          hatch build
+          echo ""
+          echo "Generated files:"
+          ls -lh dist/
 
-    - name: GH release
-      uses: softprops/action-gh-release@v2
-      with:
-        body: >
-          Please see
-          https://github.com/${{ github.repository }}/blob/${{ github.ref_name }}/CHANGELOG.md
-          for the full release notes.
-        draft: false
-        prerelease: false
+      # Store an artifact of the build to use in the publish step below
+      - name: Store the distribution packages
+        uses: actions/upload-artifact@v4
+        with:
+          name: python-package-distributions
+          path: dist/
+  publish:
+    name: >-
+      Publish Python 🐍 distribution 📦 to PyPI
+    # Modify the repo name below to be your project's repo name.
+    if: github.repository_owner == 'pyopensci'
+    needs: 
+      - build
+    runs-on: ubuntu-latest
+    # Environment required here for trusted publisher
+    environment:
+      name: pypi
+      # Modify the url to be the name of your package
+      url: https://pypi.org/p/yourPackage
+    permissions:
+      id-token: write  # this permission is mandatory for PyPI publishing
+    steps:
+      - name: Download dists
+        uses: actions/download-artifact@v4
+        with:
+          name: python-package-distributions
+          path: dist/
```

To place the source distribution and the wheel next to each other, this change would be required:

```suggestion
      - name: Download dists
        uses: actions/download-artifact@v4
        with:
          name: python-package-distributions
          path: dist/
          merge-multiple: true
```

Relevant docs: https://github.com/actions/download-artifact/tree/v4/?tab=readme-ov-file#inputs

---

Hey, @agriyakhetarpal. Please tell me more about placing them next to each other. i see in the docs that GitHub created directories for each archive. BUT, it seems to be ok for without. 

I may be missing something!

---

The idea is that the `download-artifact@v4` action (differing in behaviour from `v3`) will download the artifact under a folder `dist/python-package-distributions/`, when where we actually want them to be placed is under `dist/` itself (i.e., `dist/mypackage-version.tar.gz` and `dist/mypackage-version-<...>.whl`). Setting `merge-multiple: true` takes care of this here.

---

Oh this is a new feature! ok great suggestion. thank you for explaining!